### ISSUE TITLE
Touch handler stop working when Start/Stop VR and back to iOS native view controller

### ISSUE THREAD

**Issue by jplongo** [googlevr/gvr-unity-sdk#780]
I've embeded Unity project in native iOS app.
Everything is working here. I can open my unity nonVR scene, back to native view with a quit button.
When my view is unity scene mode I can toggle from VR to nonVR view.

The problem is if in my unity view I start&stop VR and quit it to go back to my native view, touch support stop working. I made some tests and it seems that GVR lock it even unity view is disabled.
This problem happened only if I've enable GVR. Everything is working fine if I just go and back from native view to unity view.

How can I fix this issue ?

Found using:
 * Google VR SDK version: 1.100.0
 * Unity version: 2017.1.2p2
 * Phone manufacturer, model, and O/S version: Apple iPhone 6s, iOS 11.1.1
 * Viewer manufacturer & model: n/c

Steps to reproduce the issue:
 1. Embed unity app in native iOS app. Toggle on/off VR mode. Back to native view controller

**Comment by rusmaxham** [comment]
Can you share a repro project for us to look at?

**Comment by jplongo** [comment]
It will be difficult for me to give you a demo project. I'm really busy at job now. I will try to send you a custom demo project as soon as possible.

Just in case here is how I quit my App from iOS:

_Config my project to use GVRSDK from Unity by using this settings:_
![vrenabled](https://user-images.githubusercontent.com/22789876/32772288-b48b461a-c925-11e7-8f7b-af9af2e1279e.png)


## 1 - With this script I Enable/Disable GVR on my project :
```
public class SwitchModeHandler : MonoBehaviour
{
    public Button VRMode;

    public GameObject ScrollList;
    public GameObject ShowScrolllist;
    public GameObject SaveButton;
    public GameObject EditorEmulator;
    public GameObject MainCamera;
    public GameObject VRCamera;
    public GameObject QuitButton;

    void Start() {
        Debug.Log("Init button");
        Button btnVR = VRMode.GetComponent<Button>();
        btnVR.onClick.AddListener(SwitchMode);
    }

    private void SwitchMode() {
        Debug.Log("Touch toggle button");
        if (VRSettings.enabled == false) {
            Debug.Log("Start VR");
            StartCoroutine(SwitchToVR());
        } else {
            Debug.Log("Stop VR");
            StartCoroutine(SwitchOutOfVr());
        }
    }

    // Turn on VR
    IEnumerator SwitchToVR() {
        CanvasVRModeIsOn(true); // Hide canvas elements used in 3D mode
        yield return null; // Wait one frame!
        VRSettings.LoadDeviceByName("cardboard");
        yield return null; // Wait one frame!
        VRSettings.enabled = true;
        yield return null; // Wait one frame!
    }

    // Turn off VR
    IEnumerator SwitchOutOfVr() {
        CanvasVRModeIsOn(false); // Restore canvas elements used in 3D mode
        yield return null; // Wait one frame!
        VRSettings.LoadDeviceByName("");
        yield return null; // Wait one frame!
        VRSettings.enabled = false;
        yield return null; // Wait one frame!
        Camera.main.ResetAspect(); // Unity bug don't provide aspect reset. So this line do the job on non VR camera )
    }

    private void CanvasVRModeIsOn(bool isVR) {
        SelectObject selectObjectScript = (SelectObject) GameObject.Find("Scripts").GetComponent<SelectObject>();
        selectObjectScript.enabled = !isVR;

        ScrollList.SetActive(false); // Don't need it even when VR is off!. Provide cleaner view mode
        ShowScrolllist.SetActive(!isVR);
        SaveButton.SetActive(!isVR);
        QuitButton.SetActive(!isVR);
        EditorEmulator.SetActive(isVR);
        MainCamera.SetActive(!isVR);
        VRCamera.SetActive(isVR);

        // Hide VR button without inactive it. Needed to catch GVR back button signal
        GameObject.Find("VRMode").transform.localScale = (isVR) ? GameObject.Find("VRMode").transform.localScale = new Vector3(0, 0, 0) : GameObject.Find("VRMode").transform.localScale = new Vector3(1, 1, 1);
    }

    // Catch GVR back button tapped. Unity consider it as Escape key pressed. Catch it and perform VR Off action =)
    void Update() {
        if (Input.GetKey(KeyCode.Escape))
            StartCoroutine(SwitchOutOfVr()); // Turn off VR
    }
}
```

## 2 - In Unity this script is used to perform quit action and back to native view:
```
public class QuitButton : MonoBehaviour
{
	public Button yourButton;

	[System.Runtime.InteropServices.DllImport("__Internal")]
	extern static public void quitUnity();

	void Start()
	{
		Button btn = yourButton.GetComponent<Button>();
		btn.onClick.AddListener(TaskOnClick);
	}

	void TaskOnClick()
	{
		#if !UNITY_IOS && !UNITY_EDITOR
		Application.Quit();
		#endif

		#if UNITY_IOS && !UNITY_EDITOR
		quitUnity();
		#endif
	}
}
```

## 3 - In IOS this method is called from unity button. In this method, I call StopUnity() and unload view to back to my native view from this method:

```
@objc func exitUnity() {
        NotificationCenter.default.removeObserver(self)
        
        let appDelegate = UIApplication.shared.delegate as! AppDelegate
        appDelegate.stopUnity()
        UnityViewController.isUnityLoaded = true
        self.unityView!.removeFromSuperview()
        appDelegate.myOrientation = .portrait
        self.presentingViewController?.dismiss(animated: true, completion: nil)
    }
```

There is no more in my project. As I tried to explain in my Issue.
- All of the code is working and is executed even if GVR is used or not. I'm sure by using debug session on Xcode
- Stopunity() pause Unity App by calling AppWillResignActive()
- But if GVR was enabled only one time during app exist, all the code is executed but touch handler don't go back to iOS Native and is stuck on Unity.
- As I tried to explain, if I set a timer and StartUnity() few second after stopping it, Unity back to screen and touch is working fine, so I'm sure app don't crash and touch just don't respond on native view...

Hope it will help before I can send you a demo project

Thanks for help :)

**Comment by jplongo** [comment]
Hi @rusmaxham,
Here is a link to sample project: [MyApp](https://drive.google.com/file/d/1QQ4GPB-GABnogO7z2KARgeMFlCAKA_PI/view?usp=sharing)

I've embed everything so just download and run on iOS 11 device

### How to:
- Open Unity with button
- Quit Unity from `X` button
- Open Unity again with button (Everything is working here)
- Enable Cardboard with top bottom cardboard  button
- Close VR by using les arrow button
- Quit Unity from `X` button
- Here you are in case where if GVR is used, view go back to native one, but handler stay on Unity

Hope you will help me. Thanks

**Comment by jiulongw** [comment]
Haven't try your demo but have you tried instead of adding GetUnityGLView() as a subview, presenting GetUnityGLViewController() as a full screen view controller?

I was trying to fix a weird problem regarding entering / exiting Unity from Swift project and found out presenting whole view controller has less issue than adding sub view, especially I believe in your case, your sub view takes the whole screen anyway...

**Comment by jplongo** [comment]
Hi @jiulongw,
Thanks for answer. I've already tried to use `UnityGetGLViewController` and presenting it from view, but the issue is the same. No touch if GVR was enabled.

Hope I will find a way to do this on my app.

Thanks

**Comment by jplongo** [comment]
Does anybody try my project and eventually know what's happen ? 
This issue is the only one on my project and drive me crazy. I really can't find a solution to solve it..

Thanks

**Comment by bnussey** [comment]
Any updates on this @jplongo ?

**Comment by bnussey** [comment]
Hi @jplongo have downloaded your project and have replicated the issue.. Although had to tweak to get it to run as you hard coded your link to `libGVRSDK.a` to `/Users/jp/Documents/Dev/iOS/Perspective/VR/GVRSDK/Libraries/libGVRSDK.a`

Did you ever resolve this?

**Comment by jplongo** [comment]
Nope unfortunately the issue is still here. No solutions for me now. 
The .a library come from pod when you build gvrsdk. So I certainly forgot to embed it in the sources.

**Comment by bnussey** [comment]
Hey @jplongo have you tried updating to latest version of `gvr-unity-sdk`? I noticed there is a recent release.

**Comment by jplongo** [comment]
Yes, I build my project with grr-sdk-unity 1.120.0 and iOS build using gvr-sdk-ios 1.100.0.
But unfortunely issue is still here :/

**Comment by franckos** [comment]
Hello @jplongo. I'm facing the exact same issue, using Unity 5.6 with GVR 1.80
Did you found a solution for your bug?

**Comment by jplongo** [comment]
Unfortunately nope. Same issue on last Unity/GVRSDK release. Bug is still here...

**Comment by mushthak** [comment]
Did anyone found a solution?

**Comment by michaelmatela** [comment]
has anyone fixed this?

This is what I have tried so far:
I commented out the startUnity() and stopUnity() in app delegate then I found out that the startUnity() works(with splash screen) but when you close the unity to go back in the iOS application(stopUnity()) that's where the problem occurs, I still have no solution about this my only clue for this problem is it seems that the applicationWillResignActive method in UnityAppController.mm is causing the problem and we need to adjust the method.
I will try to look for a way to close the unity instead of pausing the activity to see what happens.
I hope we found the solution for this.
